Fix Slack file upload and unfurling issues (resolves #121672)

jsuar opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR addresses the file upload and unfurling issues in the Home Assistant Slack integration, as described in #121672. Specifically, it:
- Switches from using WebClient methods directly to utilizing AsyncWebClient where asynchronous await operations are required. This prevents TypeError: Incompatible types in "await" errors and ensures proper asynchronous handling of Slack API requests.
- Improves the mocking and testing of Slack API calls (e.g., auth.test, conversations_list) in the integration’s tests. This resolves the AssertionError: No mock registered for POST https://slack.com/api/auth.test errors and ensures the test suite runs cleanly.
- Improves error handling and logging, making it easier to diagnose issues with Slack file uploads and unfurling in the future.

In [April 2024, Slack announced adjustments to their file upload methods](https://api.slack.com/changelog/2024-04-a-better-way-to-upload-files-is-here-to-stay), making it clear that a newer, more stable approach to uploading and managing files is here to stay. This clarification deprecates older workflows and reinforces that developers should rely on updated, officially supported methods (like files.remote.* and files.upload.v2) for future compatibility and consistent behavior. By aligning the Home Assistant Slack integration with these updated methods, we ensure the integration remains robust and compatible with Slack’s evolving platform standards.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #121672

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.
  - https://github.com/home-assistant/core/pull/132486#pullrequestreview-2485097639
  - https://github.com/home-assistant/core/pull/132079#pullrequestreview-2485564714
  - https://github.com/home-assistant/core/pull/132571#pullrequestreview-2489455408 
  - https://github.com/home-assistant/core/pull/131780#pullrequestreview-2489471574
 
[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jsuar committed:
Update Slack integration to use slack_sdk, add tests, and adjust codeowners

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

FletcherAU commented:
It might be worth adding a note to the docs regarding the preview image resizing? It's not necessarily intuitive unless you're familiar with Slack's preview requirements.

FletcherAU commented:
(That's definitely a  "has this been considered" more than a "you should do this" by the way)

FletcherAU commented on homeassistant/components/slack/notify.py:
```diff
@@ -208,24 +251,167 @@ async def _async_send_remote_file_message(
-            "title": title or filename,
-            "token": self._client.token,
-        }
+        # Step 3: Resize the image to meet the minimum dimension requirements
+        # https://api.slack.com/methods/files.remote.add#markdown
+        try:
+            data = await resp.read()  # Read the bytes from the StreamReader
+            # Open the image and convert to RGB if necessary
+            image = Image.open(BytesIO(data))
+            if image.mode in ("RGBA", "P"):  # Convert if image has alpha channel
+                image = image.convert("RGB")
+
+            # Get current dimensions
+            width, height = image.size
+
+            # Determine the scaling factor to meet minimum dimensions
+            scale_w = 800 / width if width < 800 else 1
+            scale_h = 400 / height if height < 400 else 1
+            scale = max(scale_w, scale_h)  # Scale up to meet both requirements
+
+            # Resize if necessary
+            if scale > 1:
+                new_width = int(width * scale)
+                new_height = int(height * scale)
+                image = image.resize((new_width, new_height), Image.LANCZOS)
+                _LOGGER.info("Image resized to %sx%s", new_width, new_height)
+            else:
+                _LOGGER.info(
+                    "Image meets the minimum dimension requirements; no resizing needed"
+                )
```
This looks like it'll fail when uploading files that aren't images.

jsuar replied:
You are correct. Thanks for the catch. I updated the code, but I need to add some test coverage for that function and the change. Hoping to get the changes pushed up in the next day or so.

jsuar replied:
Pushed up some changes to account for non-image files. Also added some more test coverage.

FletcherAU reviewed (changes requested):
Not every file upload has a generatable preview image. (The preview image is also optional within the Slack API)

jsuar committed:
Enhance Slack image upload handling: read image data on success, handle non-image files, and improve error logging

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar commented:
> It might be worth adding a note to the docs regarding the preview image resizing? It's not necessarily intuitive unless you're familiar with Slack's preview requirements.
>(That's definitely a "has this been considered" more than a "you should do this" by the way)

This is a good call out. I will review the docs and suggest possible updates

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar commented:
@FletcherAU I opened a minor documentation PR (https://github.com/home-assistant/home-assistant.io/pull/36667/) as well so I think this PR is ready for another review. Thank you! 🙇

jsuar commented:
@FletcherAU I've given this PR some further thought. If you'd prefer, I can move the file resizing code into a separate PR. This would allow this PR to focus entirely on addressing the API deprecation issue, while the file resizing and other related improvements could be handled in a future PR. Let me know what you think! 🙇

FletcherAU commented:
That might be a good idea :) (Though the reason I haven't taken a look yet is unrelated, just time poor)

jsuar committed:
Simplify remote file share

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar commented:
@FletcherAU I’ve updated the remote file upload logic to align more closely with the original implementation. Specifically, I’ve removed the image preview logic and file type checks for now. While I believe these features could be beneficial, I’ll address them in a separate PR to keep this one focused. The current changes primarily involve updates to the client and session handling. Please let me know if you spot any issues or have any concerns. Thank you!

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar commented:
@FletcherAU If it helps at all, I could reduce the PR further. At this point, having it functioning would be a win for users. Thanks for all the support 🙇

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar committed:
Merge branch 'dev' into update-slack-file-upload-issue-121672

jsuar commented:
Closing this PR in lieu of https://github.com/home-assistant/core/pull/135818 which is smaller in scope and has some newer test changes that were missed in this PR
